src/python/pants/python/python_setup.py

```diff
@@ -274,17 +279,38 @@ def get_pex_python_paths():
             return []
 
     @staticmethod
-    def get_pyenv_paths(pyenv_root_func=None):
+    def get_pyenv_paths(pyenv_root_func=None, pyenv_local=False):
         """Returns a list of paths to Python interpreters managed by pyenv.
 
         :param pyenv_root_func: A no-arg function that returns the pyenv root. Defaults to
                                 running `pyenv root`, but can be overridden for testing.
+        :param bool pyenv_local: If True, only use the interpreter specified by
+                                 '.python-version' file under `build_root`. Default to False.
         """
         pyenv_root_func = pyenv_root_func or get_pyenv_root
         pyenv_root = pyenv_root_func()
         if pyenv_root is None:
             return []
         versions_dir = os.path.join(pyenv_root, "versions")
+
+        if pyenv_local:
+
+            local_version_file = os.path.join(get_buildroot(), ".python-version")
+
+            try:
+                with open(local_version_file, "r") as f:
+                    local_version = f.readline().strip()
+                path = os.path.join(versions_dir, local_version, "bin")
+                if os.path.isdir(path):
+                    return [path]
+
+                logger.info(f"No python version {local_version} installed in pyenv.")
+
+            except IOError:
+                logger.info(f"No {local_version_file} found.")
+
+            logger.info('Fallback to use "<PYENV>" as interpreter search path')
```

Comment:
Have you used `pathlib.Path` before? See https://treyhunner.com/2018/12/why-you-should-be-using-pathlib/ for why it's generally much nicer to work with.

This code all works, but can be refactored thanks to `Path` as follows:

```python
if pyenv_local:
  local_version_file = Path(get_buildroot(), ".python-version")
  if not local_version_file.exists():
    logger.info("No `.python-version` file found in the build root, but <PYENV_LOCAL> was set in `--python-setup-interpreter-constraints`.")
    return []
  local_version = local_version_file.read_text().strip()
  path = Path(versions_dir, local_version, "bin")
  if path.is_dir():
     return [path]
  return []

paths = []
for version in sorted(...)
```

Two behavior changes in this snippet:
1) Don't fallback to `<PYENV>` if `<PYENV_LOCAL>` falls, as that is surprising behavior. We should only use `<PYENV>` if explicitly specified. So, we early return.
2) Tweaks the exception message to explain _why_ we even care if you have a `.python-version` file in the first place.

(You'll need to run `./v2 --changed-since=HEAD fmt` to format the above snippet.)

Reply:
yea, I use `pathlib` now, just see Pants code using `os.path`, so following the style.

cool idea to refactoring it by `pathlib`.

Just one thing, I guess we need to change `path` to `str` type before returning ?

Reply:
>  just see Pants code using os.path, so following the style.

Ah, gotcha. We used to exclusively use `os.path` because Pants didn't use Python 3 to run until about a year ago. We've been slowly porting to `pathlib` but there's still lots of `os.path` around.

> Just one thing, I guess we need to change path to str type before returning ? 

Ah, yes, true. You can either do `str(path)` or `path.as_posix()` to convert back to a string. We haven't settled on a particular convention - they do the same thing.